src/lib/x509/x509_ext.cpp

```diff
@@ -847,6 +856,518 @@ void TNAuthList::decode_inner(const std::vector<uint8_t>& in) {
    }
 }
 
+std::vector<uint8_t> ASBlocks::encode_inner() const {
+   std::vector<uint8_t> output;
+   DER_Encoder(output).encode(this->m_as_identifiers);
+   return output;
+}
+
+void ASBlocks::decode_inner(const std::vector<uint8_t>& in) {
+   BER_Decoder(in).decode(m_as_identifiers).verify_end();
+}
+
+void ASBlocks::ASIdentifiers::encode_into(Botan::DER_Encoder& into) const {
+   into.start_sequence();
+
+   if(m_asnum.has_value()) {
+      into.start_explicit(0);
+      into.encode(this->m_asnum.value());
+      into.end_explicit();
+   }
+
+   if(m_rdi.has_value()) {
+      into.start_explicit(1);
+      into.encode(this->m_rdi.value());
+      into.end_explicit();
+   }
+
+   into.end_cons();
+}
+
+void ASBlocks::ASIdentifiers::decode_from(Botan::BER_Decoder& from) {
+   BER_Object obj = from.get_next_object();
```

Comment:
I think this can be done as (untested)

```
from.start_sequence()
 .decode_optional(m_asnum, ASN1_Type(0), ASN1_Class::ExplicitContextSpecific);
 .decode_optional(m_rdi, ASN1_Type(1), ASN1_Class::ExplicitContextSpecific);
.end_cons()
```

If it's something much harder than that it suggests `BER_Decoder` needs a helper